osquery/tables/system/linux/disk_encryption.cpp

```diff
@@ -110,14 +115,26 @@ QueryData genFDEStatus(QueryContext& context) {
   }
 
   std::map<std::string, Row> encrypted_rows;
-  auto block_devices = SQL::selectAllFrom("block_devices");
-  for (const auto& row : block_devices) {
-    const auto name = (row.count("name") > 0) ? row.at("name") : "";
-    const auto uuid = (row.count("uuid") > 0) ? row.at("uuid") : "";
-    const auto parent_name = (row.count("parent") > 0 ? row.at("parent") : "");
-    genFDEStatusForBlockDevice(
-        name, uuid, parent_name, encrypted_rows, results);
+
+  if (context.constraints.count("name") > 0 &&
```

Comment:
if you want to optimize:
```
auto constraint = context.constraints.find("name");
if (constraint != context.constraints.end() && constraint.exists(EQUALS)) {
```

and then on line 22 `for (const auto& name = constraint.getAll(EQUALS)) {`

Reply:
Sure! I tried to implement that. See if CI compiles ;)

Reply:
This didn't compile. I dug around the source code and cobbled together something

Reply:
That didn't compile because find to a `std::map` returns an iterator which is a pointer to a `std::pair<key, value>`, so to access the value
```
auto constraint_it = context.constraints.find("name");
if(constraint_it != context.constraints.end()) {
  const auto& constraint  = constraint_it->second;
  [...] // special code for the constraint case
}

// common code
```

Also, accessing directly via `[]` it's less than ideal; we use this pattern very often but it's often incorrect, because with a `std::map` if the key is not there, a default constructed value with that key will be created into the map unnecessarily.
If the underlying data structure changes, that may instead lead to a crash.

Always prefer using find + iterator unless you are absolutely sure that the element is there.

Reply:
Reading more of our code, in some places we do something like:
```
if(context.hasConstraint("name", EQUALS) {
  for (const auto& name : context.constraints["name"].getAll(EQUALS)) {
    ...
  }
}
```

Is this a reasonable pattern?

Reply:
I guess that ends up doing a thing twice.

Reply:
It's better and safe; we need a better API though it seems :).

That is doing twice the amount of work it should be necessary, because it's accessing the map twice (unnecessarily) and also loops almost twice the constraints to find the ones matching the operation type, once for the if and once for returning the constraints.

The find + iterator + getAll on the result from the iterator it's the most efficient one and does only what's needed.

Reply:
I can't make it work. :(

I have
```
  auto constraint_it = context.constraints.find("name");
  if(constraint_it != context.constraints.end()) {
    VLOG(1) << "SEPH! Got Iterator";
    const auto uuid(""), parent_name("");
    const auto& constraint  = constraint_it->second;
    for (const auto& name : constraint.getAll(EQUALS)) {
      VLOG(1) << "SEPH! Got constraints. Checking " << name;
      runSelectAll = false;
      genFDEStatusForBlockDevice(name, uuid, parent_name, encrypted_rows, results);
    }
  }
```

The first `VLOG` always triggers, with or without a where clause. And that second `VLOG` never triggers.

Reply:
Oh good, at least partially a user error. I'd failed to rebuild enough with the new spec file.

The first `VLOG` always triggers, but now the second one does not. I'll push up some cleaner code momentarily. 

I would say we need better helpers/idioms here. :\